cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/ProgramRunDispatcherTask.java

```diff
@@ -0,0 +1,157 @@
+    if (Objects.isNull(programController)) {
+      String msg = String.format("Unable to dispatch Program %s with runid %s",
+                                 programRunDispatcherInfo.getProgramDescriptor().getProgramId(),
+                                 programRunDispatcherInfo.getRunId());
+      throw Throwables.propagate(new Throwable(msg));
+    }
+    RemoteProgramRunDispatcherResponse response = new RemoteProgramRunDispatcherResponse(true);
+    String result = GSON.toJson(response);
+    context.writeResult(result.getBytes(StandardCharsets.UTF_8));
+    executorService.submit(programRunDispatcherInfo.getCleanUpTask().get());
+  }
+
+  /**
+   * Closes this stream and releases any system resources associated
+   * with it. If the stream is already closed then invoking this
+   * method has no effect.
+   *
+   * <p> As noted in {@link AutoCloseable#close()}, cases where the
+   * close may fail require careful attention. It is strongly advised
+   * to relinquish the underlying resources and to internally
+   * <em>mark</em> the {@code Closeable} as closed, prior to throwing
+   * the {@code IOException}.
+   *
+   * @throws IOException if an I/O error occurs
+   */
+  @Override
+  public void close() throws IOException {
+    executorService.shutdown();
+  }
+
+  private static class DispatchTaskRunner {
+
+    private final CConfiguration cConf;
+    private final ProgramRunnerFactory programRunnerFactory;
+    private final ConfiguratorFactory configuratorFactory;
+    private final Impersonator impersonator;
+    private final LocationFactory locationFactory;
+    private final ArtifactRepository artifactRepository;
+    private final ProgramRunnerFactory remoteProgramRunnerFactory;
+    private final RemoteClientFactory remoteClientFactory;
+
+    @Inject
+    DispatchTaskRunner(CConfiguration cConf, ProgramRunnerFactory programRunnerFactory,
+                       ConfiguratorFactory configuratorFactory, Impersonator impersonator,
+                       LocationFactory locationFactory,
+                       @Named(AppFabricServiceRuntimeModule.NOAUTH_ARTIFACT_REPO) ArtifactRepository artifactRepository,
+                       @RemoteExecution ProgramRunnerFactory remoteProgramRunnerFactory,
+                       RemoteClientFactory remoteClientFactory) {
+      this.cConf = cConf;
+      this.programRunnerFactory = programRunnerFactory;
+      this.configuratorFactory = configuratorFactory;
+      this.impersonator = impersonator;
+      this.locationFactory = locationFactory;
+      this.artifactRepository = artifactRepository;
+      this.remoteProgramRunnerFactory = remoteProgramRunnerFactory;
+      this.remoteClientFactory = remoteClientFactory;
+    }
+
+    public ProgramController dispatch(ProgramRunDispatcherInfo programRunDispatcherInfo) throws Exception {
+      InMemoryProgramRunDispatcher dispatcher =
+        new InMemoryProgramRunDispatcher(cConf, programRunnerFactory, configuratorFactory, impersonator,
```

Comment:
Yeah that's correct. I've made the changes to use the same object from injector.

Reply:
We were doing it in this manner because previously we were getting the dispatcher instance via factory, so instead of accessing factory in system pod, I thought to directly create an instance of in memory dispatcher. But now since we're creating the dispatcher bindings in app fabric module, those will be available here.